File: src/utils/cache.ts
/**
 * Cache abstraction with in-memory and Redis adapters
 */

import { createClient, RedisClientType } from 'redis';
import { Logger, defaultLogger } from './logger';
import { CacheError } from './errors';

export interface CacheOptions {
  ttl?: number; // Time to live in seconds
}

export interface CacheAdapter {
  get<T>(key: string): Promise<T | null>;
  set<T>(key: string, value: T, options?: CacheOptions): Promise<void>;
  delete(key: string): Promise<void>;
  clear(): Promise<void>;
  has(key: string): Promise<boolean>;
}

/**
 * In-memory cache entry
 */
interface CacheEntry<T> {
  value: T;
  expiresAt: number | null;
}

/**
 * In-memory cache implementation with TTL support
 */
export class InMemoryCache implements CacheAdapter {
  private cache: Map<string, CacheEntry<any>>;
  private cleanupInterval: NodeJS.Timeout | null;
  private logger: Logger;

  constructor(logger?: Logger, cleanupIntervalMs: number = 60000) {
    this.cache = new Map();
    this.logger = logger || defaultLogger.child('InMemoryCache');

    // Periodic cleanup of expired entries
    this.cleanupInterval = setInterval(() => {
      this.cleanupExpired();
    }, cleanupIntervalMs);
  }

  /**
   * Get value from cache
   */
  async get<T>(key: string): Promise<T | null> {
    const entry = this.cache.get(key);

    if (!entry) {
      return null;
    }

    // Check if expired
    if (entry.expiresAt && Date.now() > entry.expiresAt) {
      this.cache.delete(key);
      return null;
    }

    return entry.value as T;
  }

  /**
   * Set value in cache
   */
  async set<T>(key: string, value: T, options?: CacheOptions): Promise<void> {
    const expiresAt = options?.ttl
      ? Date.now() + (options.ttl * 1000)
      : null;

    this.cache.set(key, {
      value,
      expiresAt,
    });
  }

  /**
   * Delete value from cache
   */
  async delete(key: string): Promise<void> {
    this.cache.delete(key);
  }

  /**
   * Clear all cache entries
   */
  async clear(): Promise<void> {
    this.cache.clear();
  }

  /**
   * Check if key exists in cache
   */
  async has(key: string): Promise<boolean> {
    const value = await this.get(key);
    return value !== null;
  }

  /**
   * Get cache size
   */
  size(): number {
    return this.cache.size;
  }

  /**
   * Clean up expired entries
   */
  private cleanupExpired(): void {
    const now = Date.now();
    let cleanedCount = 0;

    for (const [key, entry] of this.cache.entries()) {
      if (entry.expiresAt && now > entry.expiresAt) {
        this.cache.delete(key);
        cleanedCount++;
      }
    }

    if (cleanedCount > 0) {
      this.logger.debug(`Cleaned up ${cleanedCount} expired cache entries`);
    }
  }

  /**
   * Stop cleanup interval
   */
  destroy(): void {
    if (this.cleanupInterval) {
      clearInterval(this.cleanupInterval);
      this.cleanupInterval = null;
    }
  }
}

/**
 * Redis cache configuration
 */
export interface RedisCacheConfig {
  url?: string;
  host?: string;
  port?: number;
  password?: string;
  db?: number;
  keyPrefix?: string;
}

/**
 * Redis cache implementation
 */
export class RedisCache implements CacheAdapter {
  private client: RedisClientType;
  private keyPrefix: string;
  private logger: Logger;
  private connected: boolean = false;

  constructor(config: RedisCacheConfig = {}, logger?: Logger) {
    this.keyPrefix = config.keyPrefix || 'experimeh:';
    this.logger = logger || defaultLogger.child('RedisCache');

    // Create Redis client
    const redisConfig: any = {};

    if (config.url) {
      redisConfig.url = config.url;
    } else {
      redisConfig.socket = {
        host: config.host || 'localhost',
        port: config.port || 6379,
      };
      if (config.password) {
        redisConfig.password = config.password;
      }
      if (config.db !== undefined) {
        redisConfig.database = config.db;
      }
    }

    this.client = createClient(redisConfig);

    // Error handling
    this.client.on('error', (err) => {
      this.logger.error('Redis client error', { error: err.message });
      this.connected = false;
    });

    this.client.on('connect', () => {
      this.logger.info('Redis client connected');
      this.connected = true;
    });

    this.client.on('disconnect', () => {
      this.logger.warn('Redis client disconnected');
      this.connected = false;
    });
  }

  /**
   * Connect to Redis
   */
  async connect(): Promise<void> {
    if (!this.connected) {
      try {
        await this.client.connect();
      } catch (error) {
        throw new CacheError('Failed to connect to Redis', {
          error: error instanceof Error ? error.message : String(error)
        });
      }
    }
  }

  /**
   * Get prefixed key
   */
  private getKey(key: string): string {
    return `${this.keyPrefix}${key}`;
  }

  /**
   * Get value from cache
   */
  async get<T>(key: string): Promise<T | null> {
    try {
      const value = await this.client.get(this.getKey(key));

      if (!value) {
        return null;
      }

      return JSON.parse(value) as T;
    } catch (error) {
      this.logger.error('Failed to get value from Redis', {
        key,
        error: error instanceof Error ? error.message : String(error)
      });
      throw new CacheError('Failed to get value from cache', { key });
    }
  }

  /**
   * Set value in cache
   */
  async set<T>(key: string, value: T, options?: CacheOptions): Promise<void> {
    try {
      const serialized = JSON.stringify(value);
      const prefixedKey = this.getKey(key);

      if (options?.ttl) {
        await this.client.setEx(prefixedKey, options.ttl, serialized);
      } else {
        await this.client.set(prefixedKey, serialized);
      }
    } catch (error) {
      this.logger.error('Failed to set value in Redis', {
        key,
        error: error instanceof Error ? error.message : String(error)
      });
      throw new CacheError('Failed to set value in cache', { key });
    }
  }

  /**
   * Delete value from cache
   */
  async delete(key: string): Promise<void> {
    try {
      await this.client.del(this.getKey(key));
    } catch (error) {
      this.logger.error('Failed to delete value from Redis', {
        key,
        error: error instanceof Error ? error.message : String(error)
      });
      throw new CacheError('Failed to delete value from cache', { key });
    }
  }

  /**
   * Clear all cache entries with prefix
   */
  async clear(): Promise<void> {
    try {
      const keys = await this.client.keys(`${this.keyPrefix}*`);

      if (keys.length > 0) {
        await this.client.del(keys);
      }
    } catch (error) {
      this.logger.error('Failed to clear Redis cache', {
        error: error instanceof Error ? error.message : String(error)
      });
      throw new CacheError('Failed to clear cache');
    }
  }

  /**
   * Check if key exists in cache
   */
  async has(key: string): Promise<boolean> {
    try {
      const exists = await this.client.exists(this.getKey(key));
      return exists > 0;
    } catch (error) {
      this.logger.error('Failed to check key existence in Redis', {
        key,
        error: error instanceof Error ? error.message : String(error)
      });
      return false;
    }
  }

  /**
   * Disconnect from Redis
   */
  async disconnect(): Promise<void> {
    if (this.connected) {
      await this.client.quit();
    }
  }
}

/**
 * Cache factory for creating cache instances
 */
export class CacheFactory {
  /**
   * Create in-memory cache
   */
  static createInMemory(logger?: Logger): InMemoryCache {
    return new InMemoryCache(logger);
  }

  /**
   * Create Redis cache
   */
  static createRedis(config?: RedisCacheConfig, logger?: Logger): RedisCache {
    return new RedisCache(config, logger);
  }

  /**
   * Create cache based on environment configuration
   */
  static create(logger?: Logger): CacheAdapter {
    const cacheType = process.env.CACHE_TYPE || 'memory';

    if (cacheType === 'redis') {
      const config: RedisCacheConfig = {
        url: process.env.REDIS_URL,
        host: process.env.REDIS_HOST,
        port: process.env.REDIS_PORT ? parseInt(process.env.REDIS_PORT) : undefined,
        password: process.env.REDIS_PASSWORD,
        db: process.env.REDIS_DB ? parseInt(process.env.REDIS_DB) : undefined,
        keyPrefix: process.env.CACHE_KEY_PREFIX,
      };

      return CacheFactory.createRedis(config, logger);
    }

    return CacheFactory.createInMemory(logger);
  }
}

File: sdks/python/experimeh/cache.py
"""Caching implementation for the Experimeh SDK."""

import json
import logging
import time
from abc import ABC, abstractmethod
from typing import Any, Dict, Optional

from .errors import CacheError

logger = logging.getLogger(__name__)


class CacheAdapter(ABC):
    """Abstract base class for cache adapters."""

    @abstractmethod
    async def get(self, key: str) -> Optional[Any]:
        """Get value from cache.

        Args:
            key: Cache key

        Returns:
            Cached value or None if not found
        """
        pass

    @abstractmethod
    async def set(self, key: str, value: Any, ttl: Optional[int] = None) -> None:
        """Set value in cache.

        Args:
            key: Cache key
            value: Value to cache
            ttl: Time to live in seconds
        """
        pass

    @abstractmethod
    async def delete(self, key: str) -> None:
        """Delete value from cache.

        Args:
            key: Cache key
        """
        pass

    @abstractmethod
    async def clear(self) -> None:
        """Clear all cached values."""
        pass

    @abstractmethod
    async def has(self, key: str) -> bool:
        """Check if key exists in cache.

        Args:
            key: Cache key

        Returns:
            True if key exists, False otherwise
        """
        pass

    def get_sync(self, key: str) -> Optional[Any]:
        """Synchronous version of get (for backwards compatibility).

        Args:
            key: Cache key

        Returns:
            Cached value or None if not found
        """
        import asyncio

        try:
            loop = asyncio.get_event_loop()
            if loop.is_running():
                # If event loop is running, create a new task
                raise CacheError("Cannot call sync method from async context")
            return loop.run_until_complete(self.get(key))
        except RuntimeError:
            # No event loop, create one
            return asyncio.run(self.get(key))

    def set_sync(self, key: str, value: Any, ttl: Optional[int] = None) -> None:
        """Synchronous version of set.

        Args:
            key: Cache key
            value: Value to cache
            ttl: Time to live in seconds
        """
        import asyncio

        try:
            loop = asyncio.get_event_loop()
            if loop.is_running():
                raise CacheError("Cannot call sync method from async context")
            loop.run_until_complete(self.set(key, value, ttl))
        except RuntimeError:
            asyncio.run(self.set(key, value, ttl))


class InMemoryCache(CacheAdapter):
    """In-memory cache implementation with TTL support."""

    def __init__(self, default_ttl: int = 300, max_size: int = 1000):
        """Initialize in-memory cache.

        Args:
            default_ttl: Default TTL in seconds
            max_size: Maximum number of items to cache
        """
        self._cache: Dict[str, tuple[Any, float]] = {}
        self._default_ttl = default_ttl
        self._max_size = max_size
        logger.debug(f"Initialized InMemoryCache with TTL={default_ttl}s, max_size={max_size}")

    async def get(self, key: str) -> Optional[Any]:
        """Get value from cache.

        Args:
            key: Cache key

        Returns:
            Cached value or None if not found or expired
        """
        if key not in self._cache:
            logger.debug(f"Cache miss: {key}")
            return None

        value, expires_at = self._cache[key]

        # Check if expired
        if expires_at and time.time() > expires_at:
            logger.debug(f"Cache expired: {key}")
            await self.delete(key)
            return None

        logger.debug(f"Cache hit: {key}")
        return value

    async def set(self, key: str, value: Any, ttl: Optional[int] = None) -> None:
        """Set value in cache.

        Args:
            key: Cache key
            value: Value to cache
            ttl: Time to live in seconds (uses default if not specified)
        """
        # Enforce max size
        if len(self._cache) >= self._max_size and key not in self._cache:
            # Remove oldest entry (simple FIFO)
            oldest_key = next(iter(self._cache))
            await self.delete(oldest_key)
            logger.debug(f"Cache evicted oldest entry: {oldest_key}")

        ttl = ttl if ttl is not None else self._default_ttl
        expires_at = time.time() + ttl if ttl > 0 else 0

        self._cache[key] = (value, expires_at)
        logger.debug(f"Cache set: {key} (TTL={ttl}s)")

    async def delete(self, key: str) -> None:
        """Delete value from cache.

        Args:
            key: Cache key
        """
        if key in self._cache:
            del self._cache[key]
            logger.debug(f"Cache deleted: {key}")

    async def clear(self) -> None:
        """Clear all cached values."""
        count = len(self._cache)
        self._cache.clear()
        logger.debug(f"Cache cleared: {count} entries removed")

    async def has(self, key: str) -> bool:
        """Check if key exists in cache.

        Args:
            key: Cache key

        Returns:
            True if key exists and not expired, False otherwise
        """
        value = await self.get(key)
        return value is not None

    def size(self) -> int:
        """Get current cache size.

        Returns:
            Number of items in cache
        """
        return len(self._cache)


class RedisCache(CacheAdapter):
    """Redis cache implementation."""

    def __init__(
        self,
        host: str = "localhost",
        port: int = 6379,
        db: int = 0,
        password: Optional[str] = None,
        key_prefix: str = "experimeh:",
        default_ttl: int = 300,
        **kwargs: Any,
    ):
        """Initialize Redis cache.

        Args:
            host: Redis host
            port: Redis port
            db: Redis database number
            password: Redis password
            key_prefix: Prefix for all cache keys
            default_ttl: Default TTL in seconds
            **kwargs: Additional redis connection parameters
        """
        self._host = host
        self._port = port
        self._db = db
        self._password = password
        self._key_prefix = key_prefix
        self._default_ttl = default_ttl
        self._redis_kwargs = kwargs
        self._redis: Optional[Any] = None
        logger.debug(
            f"Initialized RedisCache with host={host}:{port}, db={db}, prefix={key_prefix}"
        )

    async def connect(self) -> None:
        """Connect to Redis server.

        Raises:
            CacheError: If connection fails
        """
        try:
            import redis.asyncio as redis
        except ImportError:
            raise CacheError(
                "redis package is required for RedisCache. Install with: pip install redis"
            )

        try:
            self._redis = redis.Redis(
                host=self._host,
                port=self._port,
                db=self._db,
                password=self._password,
                decode_responses=True,
                **self._redis_kwargs,
            )
            # Test connection
            await self._redis.ping()
            logger.info(f"Connected to Redis at {self._host}:{self._port}")
        except Exception as e:
            raise CacheError(f"Failed to connect to Redis: {str(e)}")

    async def disconnect(self) -> None:
        """Disconnect from Redis server."""
        if self._redis:
            await self._redis.close()
            self._redis = None
            logger.info("Disconnected from Redis")

    def _ensure_connected(self) -> None:
        """Ensure Redis connection is established.

        Raises:
            CacheError: If not connected
        """
        if not self._redis:
            raise CacheError("Redis not connected. Call connect() first.")

    def _make_key(self, key: str) -> str:
        """Create Redis key with prefix.

        Args:
            key: Cache key

        Returns:
            Prefixed key
        """
        return f"{self._key_prefix}{key}"

    async def get(self, key: str) -> Optional[Any]:
        """Get value from cache.

        Args:
            key: Cache key

        Returns:
            Cached value or None if not found
        """
        self._ensure_connected()
        try:
            redis_key = self._make_key(key)
            value = await self._redis.get(redis_key)

            if value is None:
                logger.debug(f"Cache miss: {key}")
                return None

            logger.debug(f"Cache hit: {key}")
            # Try to parse as JSON
            try:
                return json.loads(value)
            except (json.JSONDecodeError, TypeError):
                return value

        except Exception as e:
            logger.error(f"Redis get error: {str(e)}")
            raise CacheError(f"Failed to get from cache: {str(e)}", operation="get")

    async def set(self, key: str, value: Any, ttl: Optional[int] = None) -> None:
        """Set value in cache.

        Args:
            key: Cache key
            value: Value to cache
            ttl: Time to live in seconds
        """
        self._ensure_connected()
        try:
            redis_key = self._make_key(key)
            ttl = ttl if ttl is not None else self._default_ttl

            # Serialize value as JSON if not string
            if not isinstance(value, str):
                value = json.dumps(value)

            if ttl > 0:
                await self._redis.setex(redis_key, ttl, value)
            else:
                await self._redis.set(redis_key, value)

            logger.debug(f"Cache set: {key} (TTL={ttl}s)")

        except Exception as e:
            logger.error(f"Redis set error: {str(e)}")
            raise CacheError(f"Failed to set in cache: {str(e)}", operation="set")

    async def delete(self, key: str) -> None:
        """Delete value from cache.

        Args:
            key: Cache key
        """
        self._ensure_connected()
        try:
            redis_key = self._make_key(key)
            await self._redis.delete(redis_key)
            logger.debug(f"Cache deleted: {key}")

        except Exception as e:
            logger.error(f"Redis delete error: {str(e)}")
            raise CacheError(f"Failed to delete from cache: {str(e)}", operation="delete")

    async def clear(self) -> None:
        """Clear all cached values with prefix."""
        self._ensure_connected()
        try:
            # Scan for keys with prefix
            pattern = f"{self._key_prefix}*"
            keys = []
            async for key in self._redis.scan_iter(match=pattern):
                keys.append(key)

            if keys:
                await self._redis.delete(*keys)
                logger.debug(f"Cache cleared: {len(keys)} entries removed")

        except Exception as e:
            logger.error(f"Redis clear error: {str(e)}")
            raise CacheError(f"Failed to clear cache: {str(e)}", operation="clear")

    async def has(self, key: str) -> bool:
        """Check if key exists in cache.

        Args:
            key: Cache key

        Returns:
            True if key exists, False otherwise
        """
        self._ensure_connected()
        try:
            redis_key = self._make_key(key)
            exists = await self._redis.exists(redis_key)
            return bool(exists)

        except Exception as e:
            logger.error(f"Redis has error: {str(e)}")
            raise CacheError(f"Failed to check cache: {str(e)}", operation="has")


class NoOpCache(CacheAdapter):
    """No-operation cache that doesn't store anything."""

    async def get(self, key: str) -> Optional[Any]:
        """Always returns None."""
        return None

    async def set(self, key: str, value: Any, ttl: Optional[int] = None) -> None:
        """Does nothing."""
        pass

    async def delete(self, key: str) -> None:
        """Does nothing."""
        pass

    async def clear(self) -> None:
        """Does nothing."""
        pass

    async def has(self, key: str) -> bool:
        """Always returns False."""
        return False
